File: ouka/Megusta.py
import math
import random
import datetime
import locale
import subprocess

"""
import sys
sys.path.append('ouka')
import ouka.Megusta as mg
r = mg.Megusta()   # supondo que sua classe esteja em megusta.py

r.rPrint("Esta eh uma mensagem de exemplo.")
r.rPrintln("Esta eh uma mensagem de exemplo.")
r.rPrintln()

nome = r.rInput("Qual eh o seu nome? ")
r.rPrintln("Olah, " + nome + "!")
print()
"""
class Megusta:

    # -------------------------
    # Program
    # -------------------------

    def rPrint(self, message):
        print(message, end="")

    def rPrintln(self, message=""):
        print(message)

    def rInput(self, prompt):
        return input(prompt)

    def rSaveFile(self, nomeArquivo, conteudo):
        try:
            with open(nomeArquivo, "w", encoding="utf-8") as f:
                f.write(conteudo)
            print(f"Arquivo {nomeArquivo} salvo com sucesso.")
        except Exception as e:
            print("Erro ao salvar o arquivo:", e)

    def rOpenFile(self, arquivo):
        try:
            with open(arquivo, "r", encoding="utf-8") as f:
                return f.read()
        except FileNotFoundError:
            print("Erro: O arquivo nao foi encontrado.")
            return ""

    def rOpenProgram(self, comando):
        try:
            subprocess.Popen(comando, shell=True)
        except Exception as e:
            print("Erro ao tentar abrir o Programa:", e)

    # -------------------------
    # String
    # -------------------------

    def strReplace(self, original, var1, var2):
        return original.replace(var1, var2)

    def strLength(self, s):
        return len(s)

    def strSubstring(self, original, v1, v2):
        return original[v1:v2]

    def strCharAt(self, s, i):
        return s[i]

    def strIndexOf(self, s, v):
        return s.find(v)

    def strLastIndexOf(self, s, v):
        return s.rfind(v)

    def strToLowerCase(self, s):
        return s.lower()

    def strToUpperCase(self, s):
        return s.upper()

    def strCompareTo(self, s1, s2):
        return (s1 > s2) - (s1 < s2)

    def strCompareToIgnoreCase(self, s1, s2):
        return (s1.lower() > s2.lower()) - (s1.lower() < s2.lower())

    def strEquals(self, s1, s2):
        return s1 == s2

    def strEqualsIgnoreCase(self, s1, s2):
        return s1.lower() == s2.lower()

    # -------------------------
    # Date
    # -------------------------

    def dateNow(self):
        return datetime.datetime.now()

    def dateDay(self):
        return self.dateNow().day

    def dateWeekDay(self):
        return self.dateNow().weekday() + 1

    def dateMonth(self):
        return self.dateNow().month

    def dateYear(self):
        return self.dateNow().year

    def dateSetWeekDay(self, ano, mes, dia):
        return datetime.date(ano, mes, dia).weekday() + 1

    def dateHour24(self):
        return self.dateNow().hour

    def dateMinute(self):
        return self.dateNow().minute

    def dateSecond(self):
        return self.dateNow().second

    # -------------------------
    # Array (List)
    # -------------------------

    def arrAddAll(self, lista, *valores):
        lista.extend(valores)

    def arrAdd(self, lista, valor, index=None):
        if index is None:
            lista.append(valor)
        else:
            lista.insert(index, valor)

    def arrSet(self, lista, i, valor):
        lista[i] = valor

    def arrGet(self, lista, i):
        return lista[i]

    def arrSize(self, lista):
        return len(lista)

    def arrRemove(self, lista, i):
        lista.pop(i)

    def arrClear(self, lista):
        lista.clear()

    def arrContains(self, lista, o):
        return o in lista

    def arrToArray(self, lista):
        return list(lista)

    def arrIndexOf(self, lista, o):
        return lista.index(o)

    def arrLastIndexOf(self, lista, o):
        return len(lista) - 1 - lista[::-1].index(o)

    # -------------------------
    # Math
    # -------------------------

    def mathInt(self, s):
        return int(s)

    def mathNum(self, s):
        return float(s)

    def mathBool(self, s):
        return s.lower() == "true"

    def mathFloor(self, n):
        return math.floor(n)

    def mathCeil(self, n):
        return math.ceil(n)

    def mathRound(self, n):
        return round(n)

    def mathDecimalFormat(self, numero, pattern):
        casas = pattern.count("0") + pattern.count("#") - 1
        return f"{numero:.{casas}f}"

    def mathNumberFormat(self, numero, language, country):
        locale.setlocale(locale.LC_ALL, f"{language}_{country}.UTF-8")
        return locale.format_string("%f", numero, grouping=True)

    def mathRandom(self):
        return random.random()

    def mathAbs(self, n):
        return abs(n)

    def mathMax(self, a, b):
        return max(a, b)

    def mathMin(self, a, b):
        return min(a, b)

    def mathMaxArr(self, *values):
        return max(values)

    def mathMinArr(self, *values):
        return min(values)

    def mathPow(self, base, exp):
        return math.pow(base, exp)

    def mathSqrt(self, n):
        return math.sqrt(n)

    mathSQRT1_2 = math.sqrt(0.5)
    mathSQRT2 = math.sqrt(2)

    def mathCbrt(self, n):
        return n ** (1/3)

    def mathSignum(self, n):
        return (n > 0) - (n < 0)

    mathPI = math.pi

    def mathConvertToRadians(self, graus):
        return math.radians(graus)

    def mathSin(self, a):
        return math.sin(a)

    def mathCos(self, a):
        return math.cos(a)

    def mathTan(self, a):
        return math.tan(a)

    def mathAsin(self, v):
        return math.asin(v)

    def mathAcos(self, v):
        return math.acos(v)

    def mathAtan(self, v):
        return math.atan(v)

    def mathSinh(self, v):
        return math.sinh(v)

    def mathCosh(self, v):
        return math.cosh(v)

    def mathTanh(self, v):
        return math.tanh(v)

    def mathAsinh(self, x):
        return math.log(x + math.sqrt(x*x + 1))

    def mathAcosh(self, x):
        return math.log(x + math.sqrt(x*x - 1))

    def mathAtanh(self, x):
        return 0.5 * math.log((1 + x) / (1 - x))

    def mathLog(self, n):
        return math.log(n)

    def mathLog10(self, n):
        return math.log10(n)

    mathE = math.e
    mathLN2 = math.log(2)
    mathLOG2E = math.log2(math.e)
    mathLN10 = math.log(10)
    mathLOG10E = math.log10(math.e)

    def mathExp(self, exp):
        return math.exp(exp)

    def mathLog2(self, n):
        return math.log2(n)

    def mathLog1p(self, v):
        return math.log1p(v)

File: ouka/Megusta.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <ctype.h>
#include <stdbool.h>
#include <stdarg.h>

//typedef enum {false=0, true=1} boolean;

#define MAX_INPUT 1024

/* =========================
   PROGRAM
========================= */
/*
Para compilar:

gcc megusta.c -o megusta -lm

-lm é necessário para usar math.h.

*/
/*
#include "megusta.h"

int main(){
    rPrintln("Hello World");
}
*/
void rPrint(const char *message, ...){
  if(message == NULL) {
    printf("\n");
  }
  else {
    va_list args;
    va_start(args, message);
    vprintf(message, args);
    va_end(args);
  }
}

void rPrintln(const char *message, ...){
  if(message == NULL) {
    printf("\n");
  }
  else {
    va_list args;
    va_start(args, message);
    vprintf(message, args);
    va_end(args);
    printf("\n");
  }
}

void rPrintlnEmpty(){
    printf("\n");
}

char *rInput(const char *prompt){
    static char buffer[MAX_INPUT];
    printf("%s", prompt);
    fgets(buffer, MAX_INPUT, stdin);
    buffer[strcspn(buffer,"\n")] = 0;
    return buffer;
}

/* =========================
   FILE
========================= */

void rSaveFile(const char *nomeArquivo, const char *conteudo){
    FILE *f = fopen(nomeArquivo,"w");

    if(!f){
        printf("Erro ao salvar o arquivo.\n");
        return;
    }

    fprintf(f,"%s",conteudo);
    fclose(f);

    printf("Arquivo %s salvo com sucesso.\n", nomeArquivo);
}

char *rOpenFile(const char *arquivo){
    FILE *f = fopen(arquivo,"r");

    if(!f){
        printf("Erro: arquivo nao encontrado.\n");
        return NULL;
    }

    char *buffer = malloc(100000);
    buffer[0] = '\0';

    char linha[1024];

    while(fgets(linha,sizeof(linha),f)){
        strcat(buffer,linha);
    }

    fclose(f);

    return buffer;
}

void rOpenProgram(const char *programa){
    system(programa);
}

/* =========================
   STRING
========================= */

int strLength(const char *s){
    return strlen(s);
}

char *strSubstring(const char *s,int inicio,int fim){
    int len = fim - inicio;
    char *sub = malloc(len+1);
    strncpy(sub,s+inicio,len);
    sub[len] = '\0';
    return sub;
}

char strCharAt(const char *s,int pos){
    return s[pos];
}

int strIndexOf(const char *s,const char *find){
    char *p = strstr(s,find);
    if(!p) return -1;
    return p - s;
}

int strLastIndexOf(const char *s,const char *find){
    char *result=NULL;
    char *p=(char*)s;

    while((p=strstr(p,find))){
        result=p;
        p++;
    }

    if(!result) return -1;
    return result-s;
}

int strEquals(const char *a,const char *b){
    return strcmp(a,b)==0;
}

int strCompareTo(const char *a,const char *b){
    return strcmp(a,b);
}

char *strToUpperCase(char *s){
    for(int i=0;s[i];i++)
        s[i]=toupper(s[i]);
    return s;
}

char *strToLowerCase(char *s){
    for(int i=0;s[i];i++)
        s[i]=tolower(s[i]);
    return s;
}

char *strReplace(const char *original, const char *old, const char *new_str){
    char *result;
    int i, cnt = 0;
    int new_len = strlen(new_str);
    int old_len = strlen(old);

    for (i = 0; original[i] != '\0'; i++) {
        if (strstr(&original[i], old) == &original[i]) {
            cnt++;
            i += old_len - 1;
        }
    }

    result = (char *)malloc(i + cnt * (new_len - old_len) + 1);
    i = 0;
    const char *orig = original;
    while (*orig) {
        if (strstr(orig, old) == orig) {
            strcpy(&result[i], new_str);
            i += new_len;
            orig += old_len;
        } else {
            result[i++] = *orig++;
        }
    }
    result[i] = '\0';
    return result;
}

int strCompareToIgnoreCase(const char *a, const char *b){
    const char *p1 = a;
    const char *p2 = b;
    while (*p1 && *p2) {
        char c1 = tolower(*p1);
        char c2 = tolower(*p2);
        if (c1 != c2) return c1 - c2;
        p1++;
        p2++;
    }
    return tolower(*p1) - tolower(*p2);
}

int strEqualsIgnoreCase(const char *a, const char *b){
    return strCompareToIgnoreCase(a, b) == 0;
}

/* =========================
   DATE
========================= */

int dateDay(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_mday;
}

int dateWeekDay(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_wday+1;
}

int dateMonth(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_mon+1;
}

int dateYear(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_year+1900;
}

int dateHour24(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_hour;
}

int dateMinute(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_min;
}

int dateSecond(){
    time_t t=time(NULL);
    struct tm *tm=localtime(&t);
    return tm->tm_sec;
}

int dateSetWeekDay(int year, int month, int day){
    if (month < 3) {
        month += 12;
        year -= 1;
    }
    int k = year % 100;
    int j = year / 100;
    int h = (day + (13 * (month + 1)) / 5 + k + k / 4 + j / 4 + 5 * j) % 7;
    return ((h + 6) % 7) + 1; // 1=Sunday, 7=Saturday
}

/* =========================
   ARRAY (Lista dinâmica)
========================= */

typedef struct{
    char **data;
    int size;
    int capacity;
} StringList;

void arrInit(StringList *list){
    list->size=0;
    list->capacity=10;
    list->data=malloc(sizeof(char*)*list->capacity);
}
void arrAddAll(StringList *list,...){
    va_list args;
    va_start(args, list);
    while(1){
        char *valor = va_arg(args, char*);
        if(valor == NULL) break;
        if(list->size >= list->capacity){
            list->capacity *= 2;
            list->data = realloc(list->data, sizeof(char*) * list->capacity);
        }
        list->data[list->size] = strdup(valor);
        list->size++;
    }
    va_end(args);
}

void arrAdd(StringList *list,const char *valor){

    if(list->size>=list->capacity){
        list->capacity*=2;
        list->data=realloc(list->data,sizeof(char*)*list->capacity);
    }

    list->data[list->size]=strdup(valor);
    list->size++;
}

char *arrGet(StringList *list,int i){
    return list->data[i];
}

void arrSet(StringList *list,int i,const char *valor){
    free(list->data[i]);
    list->data[i]=strdup(valor);
}

int arrSize(StringList *list){
    return list->size;
}

void arrRemove(StringList *list,int i){
    free(list->data[i]);

    for(int j=i;j<list->size-1;j++)
        list->data[j]=list->data[j+1];

    list->size--;
}

void arrClear(StringList *list){
    for(int i=0;i<list->size;i++)
        free(list->data[i]);

    list->size=0;
}

int arrContains(StringList *list, const char *valor){
    for(int i = 0; i < list->size; i++){
        if(strcmp(list->data[i], valor) == 0) return 1;
    }
    return 0;
}

char **arrToArray(StringList *list){
    return list->data;
}

int arrIndexOf(StringList *list, const char *valor){
    for(int i = 0; i < list->size; i++){
        if(strcmp(list->data[i], valor) == 0) return i;
    }
    return -1;
}

int arrLastIndexOf(StringList *list, const char *valor){
    for(int i = list->size - 1; i >= 0; i--){
        if(strcmp(list->data[i], valor) == 0) return i;
    }
    return -1;
}

/* =========================
   MATH
========================= */

long mathInt(const char *s){
    return atol(s);
}

double mathNum(const char *s){
    return atof(s);
}

double mathFloor(double x){
    return floor(x);
}

double mathCeil(double x){
    return ceil(x);
}

double mathRound(double x){
    return round(x);
}

double mathRandom(){
    return (double)rand()/RAND_MAX;
}

double mathAbs(double x){
    return fabs(x);
}

double mathMax(double a,double b){
    return fmax(a,b);
}

double mathMin(double a,double b){
    return fmin(a,b);
}

double mathPow(double b,double e){
    return pow(b,e);
}

double mathSqrt(double n){
    return sqrt(n);
}

int mathBool(const char *s){
    return strcmp(s, "true") == 0 || strcmp(s, "1") == 0;
}

char *mathDecimalFormat(double num, const char *pattern){
    static char buf[256];
    if(strcmp(pattern, "#.###") == 0){
        sprintf(buf, "%.3f", num);
        return buf;
    }
    sprintf(buf, "%f", num);
    return buf;
}

char *mathNumberFormat(double num, const char *locale, const char *country){
    static char buf[256];
    sprintf(buf, "%.2f", num);
    return buf;
}

double mathMaxArr(double a, double b, double c, double d){
    return fmax(fmax(a, b), fmax(c, d));
}

double mathMinArr(double a, double b, double c, double d){
    return fmin(fmin(a, b), fmin(c, d));
}

#define mathSQRT1_2 0.7071067811865476
#define mathSQRT2 1.4142135623730951

double mathCbrt(double x){
    return cbrt(x);
}

double mathSignum(double x){
    if (x > 0) return 1.0;
    if (x < 0) return -1.0;
    return 0.0;
}

#define mathPI 3.141592653589793

double mathConvertToRadians(double degrees){
    return degrees * mathPI / 180.0;
}

double mathSin(double x){
    return sin(x);
}

double mathCos(double x){
    return cos(x);
}

double mathTan(double x){
    return tan(x);
}

double mathAsin(double x){
    return asin(x);
}

double mathAcos(double x){
    return acos(x);
}

double mathAtan(double x){
    return atan(x);
}

double mathSinh(double x){
    return sinh(x);
}

double mathCosh(double x){
    return cosh(x);
}

double mathTanh(double x){
    return tanh(x);
}

double mathAsinh(double x){
    return asinh(x);
}

double mathAcosh(double x){
    return acosh(x);
}

double mathAtanh(double x){
    return atanh(x);
}

double mathLog(double x){
    return log(x);
}

double mathLog10(double x){
    return log10(x);
}

#define mathE 2.718281828459045
#define mathLN2 0.6931471805599453
#define mathLOG2E 1.4426950408889634
#define mathLN10 2.302585092994046
#define mathLOG10E 0.4342944819032518

double mathExp(double x){
    return exp(x);
}

double mathLog2(double x){
    return log2(x);
}

double mathLog1p(double x){
    return log1p(x);
}

void arrInsert(StringList *list, int index, const char *valor){
    if(index < 0) index = 0;
    if(index > list->size) index = list->size;
    if(list->size >= list->capacity){
        list->capacity *= 2;
        list->data = realloc(list->data, sizeof(char*) * list->capacity);
    }
    for(int j = list->size; j > index; j--){
        list->data[j] = list->data[j - 1];
    }
    list->data[index] = strdup(valor);
    list->size++;
}